services/agents/registry.go

```diff
@@ -456,6 +459,65 @@ func (r *Registry) SendSetStateRequest(ctx context.Context, pmmAgentID string) {
 	l.Infof("SetState response: %+v.", resp)
 }
 
+// CheckConnectionToService sends request to pmm-agent to check connection to service.
+func (r *Registry) CheckConnectionToService(ctx context.Context, service *models.Service, agent *models.Agent) error {
+	// TODO: extract to a separate struct to keep Single Responsibility principles.
+	l := logger.Get(ctx)
+	start := time.Now()
+	defer func() {
+		if dur := time.Since(start); dur > time.Second {
+			l.Warnf("CheckConnectionToService took %s.", dur)
+		}
+	}()
+
+	pmmAgentID := pointer.GetString(agent.PMMAgentID)
+	pmmAgent, err := r.Get(ctx, pmmAgentID)
+	if err != nil {
+		l.Infof("CheckConnectionToService: pmm-agent with ID %q is not currently connected.", pmmAgentID)
+		return err
+	}
+	var request *agentpb.CheckConnectionRequest
+
+	switch service.ServiceType {
+	case models.MySQLServiceType:
+		request = &agentpb.CheckConnectionRequest{
+			Type: inventorypb.ServiceType_MYSQL_SERVICE,
+			Dsn:  mysqlDSN(service, agent),
+		}
+	case models.PostgreSQLServiceType:
+		request = &agentpb.CheckConnectionRequest{
+			Type: inventorypb.ServiceType_POSTGRESQL_SERVICE,
+			Dsn:  postgresqlDSN(service, agent),
+		}
+	case models.MongoDBServiceType:
+		request = &agentpb.CheckConnectionRequest{
+			Type: inventorypb.ServiceType_MONGODB_SERVICE,
+			Dsn:  mongoDSN(service, agent),
+		}
+	default:
+		l.Panicf("unhandled Service type %s", service.ServiceType)
+	}
+
+	l.Infof("CheckConnectionRequest: %+v.", request)
+	resp := pmmAgent.channel.SendRequest(request)
+	l.Infof("CheckConnection response: %+v.", resp)
+	checkConnectionResponse := resp.(*agentpb.CheckConnectionResponse)
+	if !checkConnectionResponse.Success {
+		return nil
+	}
+	return status.Error(codes.FailedPrecondition, checkConnectionResponse.Error)
+}
+
+func (r *Registry) Get(ctx context.Context, pmmAgentID string) (*agentInfo, error) {
+	r.rw.RLock()
+	pmmAgent := r.agents[pmmAgentID]
+	r.rw.RUnlock()
+	if pmmAgent == nil {
+		return nil, status.Error(codes.Unimplemented, fmt.Sprintf("pmm-agent with ID %q is not currently connected.", pmmAgentID))
```

Comment:
I think, `codes.Unimplemented` should be changed to something else. Maybe to `codes.Internal`?